File: leetcode/2300_Successful_Pairs_of_Spells_and_Potions/Solution.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);

        int[] ans = new int[spells.length];
        for(int i = 0; i < spells.length; i++) {
            int l = 0, r = potions.length - 1;
            
            int minIdx = -99;
            while(l <= r) {
                int mid = (l + r) / 2;

                if((long)spells[i] * potions[mid] >= success) {
                    minIdx = mid;
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            }

            if(minIdx == -99) {
                ans[i] = 0;
            } else {
                ans[i] = potions.length - minIdx;
            }
        }

        return ans;
    }
}

File: leetcode/2300_Successful_Pairs_of_Spells_and_Potions/solution.py
class Solution:
    def successfulPairs(
        self, spells: List[int], potions: List[int], success: int
    ) -> List[int]:
        potions.sort()

        ans = []
        for i in range(len(spells)):
            l = 0
            r = len(potions) - 1
            min_successful_idx = -1

            while l <= r:
                m = l + (r - l) // 2

                if potions[m] * spells[i] >= success:
                    min_successful_idx = m
                    r = m - 1
                else:
                    l = m + 1

            if min_successful_idx >= 0:
                ans.append(len(potions) - min_successful_idx)
            else:
                ans.append(0)

        return ans
